### ISSUE TITLE
Failed to set remote answer sdp: Called in wrong state: stable

### ISSUE THREAD

**Issue by sm2017** [flutter-webrtc/flutter-webrtc#539]
In the `onRenegotiationNeeded` I `createOffer` and `setLocalDescription` then send it to signaling server, on the remote peer, I call `setRemoteDescription` the `createAnswer` and call `setLocalDescription` with answer

After calling `setLocalDescription` I get the error 
> Failed to set remote answer sdp: Called in wrong state: stable

I do exactly same in javascript in web client, it works and I have no error, It may be a but in flutter-webrtc , I am using latest version

**Comment by cabdirashiid-developer** [comment]
On the remote peer check peer connection signaling state then call your answer if signaling state has remote offer. 

like so: 
` _peerConnection.onSignalingState = (state) {
        if (state == RTCSignalingState.RTCSignalingStateHaveRemoteOffer) {
            // answer and set local description here
          }
      };`

**Comment by cabdirashiid-developer** [comment]
On the remote peer check peer connection signaling state then call your answer if signaling state has remote offer.

like so:
```
      _peerConnection.onSignalingState = (state) {
        if (state == RTCSignalingState.RTCSignalingStateHaveRemoteOffer) {
            // answer here
          }
      };
```

**Comment by sm2017** [comment]
> on the remote peer, I call setRemoteDescription the createAnswer and call setLocalDescription with answer

As I await for `setRemoteDescription`  with remote offer, I expect that the state will switched to `RTCSignalingState.RTCSignalingStateHaveRemoteOffer`, So what's wrong?

**Comment by cabdirashiid-developer** [comment]
> > on the remote peer, I call setRemoteDescription the createAnswer and call setLocalDescription with answer
> 
> As I await for `setRemoteDescription` with remote offer, I expect that the state will switched to `RTCSignalingState.RTCSignalingStateHaveRemoteOffer`, So what's wrong?

I didn't understand. On local or remote peer. BTW I had Webrtc working on native android but I'm struggling getting it to work on flutter. I posted an issue.

**Comment by sm2017** [comment]
On the remote peer after receiving the negotiation offer, I call `setRemoteDescription` so I expect that the state will switched to `RTCSignalingState.RTCSignalingStateHaveRemoteOffer`

**Comment by cabdirashiid-developer** [comment]
I'm a bit lost myself. I'll update you if I get it to work. Meanwhile here's [an explanation of signaling states](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/signalingState):

**Comment by sm2017** [comment]
@cabdirashiid-developer My video and voice call application works with flutter webttc, without problem, I see your issue too and found that its completely different with my application, My problem is just re-negotiation

**Comment by cabdirashiid-developer** [comment]
Yeah I think I figured it out. The local peer wasn't adding remote candidates or setting remote description. I was expecting the database values to trigger it when there was a change but it wasn't, I'm working on it now.

I went minimal with my code.

**Comment by feyisayo-dev** [comment]
I am having same issue, I dont even know where it all went wrong
i am trying client side javascript sm2017 can u help since urs is working
to try to battle it i used
function setRemoteDescription(description) {
    if (peerConnection.signalingState === 'stable' || peerConnection.signalingState === 'have-local-offer') {
        // If the signaling state is 'stable' or 'have-local-offer', set the remote description directly
        peerConnection.setRemoteDescription(description)
            .then(function () {
                // Check if there are any pending ICE candidates to be added
                if (pendingCandidates.length > 0) {
                    pendingCandidates.forEach(function (candidate) {
                        peerConnection.addIceCandidate(candidate)
                            .catch(function (error) {
                                console.log('Error handling pending ICE candidate:', error);
                            });
                    });
                    // Clear the pending candidates array
                    pendingCandidates = [];
                }
            })
            .catch(function (error) {
                console.log('Error handling call answer:', error);
            });
    } else {
        peerConnection.addEventListener('signalingstatechange', function () {
            if (peerConnection.signalingState === 'stable' || peerConnection.signalingState === 'have-local-offer') {
                peerConnection.setRemoteDescription(description)
                    .then(function () {
                        // Check if there are any pending ICE candidates to be added
                        if (pendingCandidates.length > 0) {
                            pendingCandidates.forEach(function (candidate) {
                                peerConnection.addIceCandidate(candidate)
                                    .catch(function (error) {
                                        console.log('Error handling pending ICE candidate:', error);
                                    });
                            });
                            // Clear the pending candidates array
                            pendingCandidates = [];
                        }
                    })
                    .catch(function (error) {
                        console.log('Error handling call answer:', error);
                    });
            }
        });
    }
}
same error
